webhook/cmd/webhook.go

```diff
@@ -0,0 +1,238 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"os"
+	"strconv"
+
+	v1beta1 "k8s.io/api/admission/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type patchOperation struct {
+	Op    string      `json:"op"`
+	Path  string      `json:"path"`
+	Value interface{} `json:"value,omitempty"`
+}
+
+func getEnv(key, fallback string) string {
+	value, exists := os.LookupEnv(key)
+	if !exists {
+		value = fallback
+	}
+	return value
+}
+
+func createPatch(pod *corev1.Pod) ([]byte, error) {
+	patch := []patchOperation{}
+	patch = append(patch, addVolume(pod)...)
+	patch = append(patch, updateContainers(pod)...)
+	patch = append(patch, addInitContainer(pod)...)
+	return json.Marshal(patch)
+}
+
+func addVolume(pod *corev1.Pod) (patch []patchOperation) {
+
+	volume := corev1.Volume{
+		Name: "hypertrace-javaagent",
+		VolumeSource: corev1.VolumeSource{
+			EmptyDir: &corev1.EmptyDirVolumeSource{},
+		},
+	}
+
+	path := "/spec/volumes"
+	var value interface{}
+
+	if len(pod.Spec.Volumes) != 0 {
+		path = path + "/-"
+		value = volume
+	} else {
+		value = []corev1.Volume{volume}
+	}
+
+	patch = append(patch, patchOperation{
+		Op:    "add",
+		Path:  path,
+		Value: value,
+	})
+
+	return patch
+}
+
+func getServiceName(pod *corev1.Pod) (serviceName string) {
+	if value, found := pod.Labels["app.kubernetes.io/name"]; found {
+		return value
+	}
+
+	if value, found := pod.Labels["app"]; found {
+		return value
+	}
+
+	return ""
+}
+
+// add volumeMount and JAVA_TOOL_OPTIONS environment variable
+func updateContainers(pod *corev1.Pod) (patch []patchOperation) {
+
+	modifiedContainers := []corev1.Container{}
+	serviceName := getServiceName(pod)
+
+	for _, container := range pod.Spec.Containers {
+		ignore := false
+		for _, envVar := range container.Env {
+			if envVar.Name == "HYPERTRACE_IGNORE_JAVAAGENT" {
```

Comment:
we should probably use a label for this.

Reply:
we needed it at container level. labels are pod specific. This is added not to inject java-agent to one of the container in multi-containers pod.

Reply:
from the UX perspective we need these features: inject agent if a lebel is present e.g. on namespace/deployment/pod. Exclude specific deployments/pods if injection is enabled on a namespace. 

We don't have to implement these right away, we just should know that it is feasible.

Reply:
We will also need to provide agent configuration in config map - see `agent-attacher-config-map` in traceableai nameslace. This CM embeds the hypertrace agent config that the agent uses.

Reply:
```
k get configmap agent-attacher-config-map -n traceableai -o json | jq '.data."agent-attacher-config.yaml"' -r                                                                                                       1 ↵ ploffay@Pavols-MBP
agent_attacher:
  logging:
    level: info
    encoding: "json"
    outputPaths:
      - stdout
    errorOutputPaths:
      - stderr
    encoderConfig:
      timeKey: "time"
      timeEncoder: "iso8601"
      levelKey: "level"
      levelEncoder: "uppercase"
      messageKey: "message"
  reporting:
    address: api-staging.traceable.ai
  kubernetes:
    auto_instrument_services: false
    namespaces:
      - label_selector:
          traceableai-instrumentation=enabled
        attach_disabled: false
        instrumentation_agent:
          use_hypertrace_agent: true
          hypertrace_agent_config:
            reporting:
              endpoint: "http://collector.traceableai:9411/api/v2/spans"
            opa:
              endpoint: "http://opa.traceableai:8181/"
            dataCapture:
              httpHeaders:
                request: true
                response: true
              httpBody:
                request: true
                response: true
              rpcMetadata:
                request: true
                response: true
              rpcBody:
                request: true
                response: true
          logging:
            level: info
            outputPaths:
              - /tmp
            errorOutputPaths:
              - /tmp
          reporting:
            address: api-staging.traceable.ai
            traces_address: "http://collector.traceableai:14268/api/traces"
          java_agent:
            jattach: true
          disable_capture_of_attributes:
            - http.request.body
            - http.request.header
            - http.response.body
            - http.response.header

    pods:
      - label_selector:
          traceableai-instrumentation=enabled
        attach_disabled: false
        instrumentation_agent:
          use_hypertrace_agent: true
          hypertrace_agent_config:
            reporting:
              endpoint: "http://collector.traceableai:9411/api/v2/spans"
            opa:
              endpoint: "http://opa.traceableai:8181/"
            dataCapture:
              httpHeaders:
                request: true
                response: true
              httpBody:
                request: true
                response: true
              rpcMetadata:
                request: true
                response: true
              rpcBody:
                request: true
                response: true
          logging:
            level: info
            outputPaths:
              - /tmp
            errorOutputPaths:
              - /tmp
          reporting:
            address: api-staging.traceable.ai
            traces_address: "http://collector.traceableai:14268/api/traces"
          java_agent:
            jattach: true
          disable_capture_of_attributes:
            - http.request.body
            - http.request.header
            - http.response.body
            - http.response.header
      - label_selector:
          traceableai-instrumentation=disabled
        attach_disabled: true

```

Reply:
configmaps are namespace scoped and to mount it on a pod, we have to make sure that the configmap exists in the namespace of the pod. Instead, we can mount this configmap on the webhook pod and copy the YAML file to the same shared volume where the java agent jar is copied.

Reply:
@pavolloffay @davexroth @mohit-a21 hypertrace javaagent needs the whole configmap or just the section under `hypertrace_agent_config`?

Reply:
just the section under `hypertrace_agent_config`. The config proto is here https://github.com/hypertrace/agent-config.